kubernetes/cache/proxy_status.go

```diff
@@ -1,31 +1,61 @@
 package cache
 
 import (
+	"context"
 	"strings"
 	"time"
 
+	"k8s.io/apimachinery/pkg/util/wait"
+
 	"github.com/kiali/kiali/kubernetes"
+	"github.com/kiali/kiali/log"
 )
 
-type (
-	ProxyStatusCache interface {
-		CheckProxyStatus() bool
-		GetPodProxyStatus(namespace, pod string) *kubernetes.ProxyStatus
-		SetProxyStatus(proxyStatus []*kubernetes.ProxyStatus)
-		RefreshProxyStatus()
-	}
-)
+type ProxyStatusCache interface {
+	GetPodProxyStatus(namespace, pod string) *kubernetes.ProxyStatus
+}
 
-func (c *kialiCacheImpl) CheckProxyStatus() bool {
-	defer c.proxyStatusLock.RUnlock()
-	c.proxyStatusLock.RLock()
-	if c.proxyStatusCreated == nil {
-		return false
-	}
-	if time.Since(*c.proxyStatusCreated) > c.tokenNamespaceDuration {
-		return false
-	}
-	return true
+// pollIstiodForProxyStatus is a long running goroutine that will periodically poll istiod for proxy status.
+// Polling stops when the stopCacheChan is closed.
+func (c *kialiCacheImpl) pollIstiodForProxyStatus() {
+	log.Debug("[Kiali Cache] Scraping istiod for proxy status")
+	go func() {
+		for {
+			select {
+			case <-c.stopCacheChan:
+				return
+			case <-time.After(c.tokenNamespaceDuration):
+				// Get the proxy status from istiod with some retries.
+				// Wrapping this so we can defer cancel.
+				func() {
+					ctx, cancel := context.WithTimeout(context.Background(), c.tokenNamespaceDuration)
+					defer cancel()
+
+					var (
+						proxyStatus []*kubernetes.ProxyStatus
+						err         error
+					)
+
+					retryErr := wait.PollImmediateUntilWithContext(ctx, time.Second*4, func(ctx context.Context) (bool, error) {
```

Comment:
The fixed 4s is probably OK because the 10s setting for tokenNamespaceDuration is probably not going to change, but if we wanted to ensure at least 1 retry, or say limit to 2 retries, then the retry time should be based on the config setting.

Reply:
Updated this to be 1/2 `tokenNamespaceDuration`